### PR TITLE
DAO-209 Load and display payout info on the Claim Details page

### PR DESCRIPTION
**Pull request by mcoetzee**
### What does this change?
- it loads payout data on the Claim Details page
- it displays the payout amount (API3 tokens), a help/info tooltip, and the transaction link for claims with status:
    - `ClaimAccepted`
    - `SettlementAccepted`
    - `DisputeResolvedWithClaimPayout`
    - `DisputeResolvedWithSettlementPayout`
- it preloads payout data on the My Claims page in order to improve the Claims Details page load

### How did you action this task?
- the `AcceptedClaim`, `AcceptedSettlement`, `ResolvedDisputeByAcceptingClaim`, `ResolvedDisputeByAcceptingSettlement` events were used to load the payout data for the claims
- to simplify the Claim Details page, the claim and the payout data is loaded together

### Screenshots
Screenshot 2022-09-26 at 15 44 06

### CODE REVIEW

File: src/contracts/network.ts [hunk 1]
```diff
@@ -37,12 +36,14 @@ export const ETHERSCAN_HOSTS: { [chainId: string]: string } = {
   42: 'https://kovan.etherscan.io',
 };
 
-export const getEtherscanTransactionUrl = (transaction: ethers.Transaction) => {
-  const host = ETHERSCAN_HOSTS[transaction.chainId.toString()];
+export const getEtherscanTransactionUrl = (chainId: number | undefined, transactionHash: string) => {
+  if (!chainId) return;
```
Comment (Siegrift) [inline src/contracts/network.ts:40]: Nit: Maybe combine these two ifs?

File: src/logic/claims/data.ts [hunk 2]
```diff
@@ -98,7 +109,8 @@ export function useUserClaimById(claimId: string) {
       setChainData(
         'Loaded claim',
         updateImmutablyCurried((state) => {
```
Comment (Siegrift) [inline src/logic/claims/data.ts:111]: Retrospectively, I really dislike this `updateImmutablyCurried` name. 
Maybe at this point we might plug in some proper state management library like zustand/redux to help with performance.

WDYT?

File: src/logic/claims/data.ts [hunk 3]
```diff
@@ -329,3 +361,64 @@ function getDisputeStatus(period: DisputePeriod): DisputeStatus {
...
+
+  const events = sortEvents([
+    ...acceptedEvents,
+    ...settlementEvents,
+    ...resolvedClaimEvents,
+    ...resolvedSettlementEvents,
+  ] as const);
+
+  const byId = events.reduce((acc, ev) => {
+    if ('clippedAmountInUsd' in ev.args) {
```
Comment (Siegrift) [inline src/logic/claims/data.ts:435]: You probably want to use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty

File: src/logic/claims/data.ts [hunk 4]
```diff
@@ -329,3 +361,64 @@ function getDisputeStatus(period: DisputePeriod): DisputeStatus {
       return 'Solved';
   }
 }
+
+async function loadClaimPayoutData(contract: ClaimsManager, params: { userAccount?: string; claimIds?: string[] }) {
+  const { userAccount = null, claimIds = null } = params;
+  const [acceptedEvents, settlementEvents, resolvedClaimEvents, resolvedSettlementEvents] = await Promise.all([
+    contract.queryFilter(
+      contract.filters.AcceptedClaim(
+        // @ts-expect-error
```
Comment (Siegrift) [inline src/logic/claims/data.ts:398]: nit: provide reason

File: src/pages/claim-details/claim-actions.tsx [hunk 5]
```diff
@@ -180,7 +180,17 @@ export default function ClaimActions(props: Props) {
               </Tooltip>
             </div>
           </div>
-          <p className={styles.actionMessage}>All done! The claim payout has been accepted.</p>
+          <p className={styles.actionMessage}>
+            All done! The claim payout has been accepted.{' '}
+            <a
```
Comment (Siegrift) [inline src/pages/claim-details/claim-actions.tsx]: Maybe use the `ExternalLink` compoent?

Comment (mcoetzee) [reply src/pages/claim-details/claim-actions.tsx]: Yeah that's the plan 👍. I've logged a separate issue to make ExternalLink easier to reuse (it's got unnecessary styling that makes it a bit of a pain): https://linear.app/api3-dao/issue/DAO-211/remove-styling-from-externallink-component-to-make-it-easier-to-reuse

As part of that task I will use ExternalLink where appropriate

File: src/logic/claims/data.ts (hunk 2 above)
Comment (mcoetzee) [reply src/logic/claims/data.ts:111]: I'm not a fan of the `updateImmutablyCurried` name either, but I don't think we need to reach for a state management library. The app is simple enough. We can probably get rid of `updateImmutablyCurried` though and tweak the `setChainData()` API to:
```ts
setChainData(
  'Loaded policies',
  (state) => {
    state.policies.userPolicyIds = result.data.ids;
    state.policies.byId = { ...state.policies.byId, ...result.data.byId };
  }
);
```

Comment (Siegrift) [reply src/logic/claims/data.ts:111]: or simply call it `updateState` ?

File: src/logic/claims/data.ts (hunk 3 above)
Comment (mcoetzee) [reply src/logic/claims/data.ts:435]: Using `in` is a common way to narrow down types in Typescript: https://www.typescriptlang.org/docs/handbook/2/narrowing.html#the-in-operator-narrowing

Comment (Siegrift) [reply src/logic/claims/data.ts:435]: Yeah, but I dislike the fact that `in` traverses the full prototype chain and in some cases results in wrong logic.

Comment (mcoetzee) [reply src/logic/claims/data.ts:435]: I think this is a matter of preference 🙂. I highly doubt we are at risk of any wrong logic here ito inherited properties.
Also, using `hasOwnProperty()` doesn't give us the type narrowing we want here:
Screenshot 2022-09-29 at 12 48 23

Lastly, there is also the Eslint rule against using `hasOwnProperty()` directly: https://eslint.org/docs/latest/rules/no-prototype-builtins

File: src/logic/claims/data.ts (hunk 2 above)
Comment (mcoetzee) [reply src/logic/claims/data.ts:111]: Hmm Immer uses `produce`, so I reckon we should probably follow their lead and call it `produceState`. My feeling is `setChainData` and `updateState` sounds too similar to each other.

File: src/logic/claims/data.ts (hunk 3 above)
Comment (Siegrift) [reply src/logic/claims/data.ts:435]: Yeah, feel free to leave it as is. Especially the narrowing is a big plus.

File: src/logic/claims/data.ts (hunk 2 above)
Comment (Siegrift) [reply src/logic/claims/data.ts:111]: I am fine with `produceState` as well.

Comment (mcoetzee) [reply src/logic/claims/data.ts:111]: FYI I will rename in https://linear.app/api3-dao/issue/DAO-214/rename-updateimmutablycurried-function

**Review by Siegrift — APPROVED** [review]
👍 LGTM

**Pull request merged**